query-projector/src/lib.rs

```diff
@@ -243,23 +487,119 @@ fn project_elements<'a, I: IntoIterator<Item = &'a Element>>(
+
+                    // We might regret using the type tag here instead of the `ValueType`.
+                    templates.push(TypedIndex::Known(i, return_type.value_type_tag()));
+                    i += 1;
+                } else {
+                    // TODO: complex aggregates.
+                    bail!(ErrorKind::NotYetImplemented("complex aggregates".into()));
+                }
+            },
         }
     }
 
-    for var in with {
-        // We need to collect these into the SQL column list, but they don't affect projection.
-        // If a variable is of a non-fixed type, also project the type tag column, so we don't
-        // accidentally unify across types when considering uniqueness!
-        let (column, name) = candidate_column(&query.cc, &var);
+    // Anything we're projecting, or that's part of an aggregate, doesn't need to be in GROUP BY.
+    //
+    // Anything used in ORDER BY (which we're given in `named_projection`)
+    // needs to be in the SQL column list so we can refer to it by name.
+    //
+    // They don't affect projection.
+    //
+    // If a variable is of a non-fixed type, also project the type tag column, so we don't
+    // accidentally unify across types when considering uniqueness!
+    // Similarly, the type tag needs to be grouped.
+    // extra_projection: extras - variables
+    for var in query.named_projection.iter() {
+        if variables.contains(var) {
+            continue;
+        }
+
+        // If it's a fixed value, we need do nothing further.
+        if query.cc.is_value_bound(&var) {
+            continue;
+        }
+
+        let (column, name) = candidate_column(&query.cc, &var)?;
         cols.push(ProjectedColumn(column, name));
-        if query.cc.known_type(&var).is_none() {
+
+        // We don't care if a column has a single _type_, we care if it has a single type _code_,
+        // because that's what we'll use if we're projecting. E.g., Long and Double.
+        // Single type implies single type code, and is cheaper, so we check that first.
+        let types = query.cc.known_type_set(&var);
+        if !types.has_unique_type_code() {
             let (type_column, type_name) = candidate_type_column(&query.cc, &var);
             cols.push(ProjectedColumn(type_column, type_name));
         }
     }
 
-    Ok((Projection::Columns(cols), templates))
+    if aggregated.is_empty() {
+        // We're done -- we never need to group unless we're aggregating.
+        return Ok(ProjectedElements {
+                      sql_projection: Projection::Columns(cols),
+                      templates,
+                      group_by: None,
+                  });
+    }
+
+    // group_by: (with + variables) - aggregated
+    let mut group_by_vars: BTreeSet<Variable> = query.with.union(&variables).cloned().collect();
+    for var in aggregated {
+        group_by_vars.remove(&var);
+    }
+
+    // We never need to group by a constant.
```

Comment:
I guess the number of aggregated vars is similar to the number of bound vars in most instances.  You didn't like the extra allocations in `difference`?